feat: add recipe viewers for canning foods

Roelymole opened the pull request:
Full support for viewing how to make canned foods in the food canner using REI/JEI and partial support for EMI (only the default items in the canned food creative tab have recipes, since EMI does not allow for dynamic recipes).

- [x] Animate the progress bars and show the whole process with a single can moving from the top slot to the bottom slot.
- [x] The canning tab doesn't currently appear in the list of all categories since there aren't any recipe data files for canning and instead the list of recipes is dynamically generated.
- [x] Add support for the move items button. *(Doesn't currently transfer an empty can to the top slot, which might be something worth adding.)*
- [ ] Use custom transfer handlers to allow the move items button in REI/JEI to transfer more than one item into a slot.
- [ ] Handle displaying more than 16 distinct items (not currently possible as the `MAX_FOOD` value is hardcoded, but necessary if we make this value configurable).

An example of the recipe viewer for a can with 16 different foods (by default this can be seen by pressing R when hovering over the item).

<img width="1920" height="1080" alt="2025-08-16_18 18 29" src="https://github.com/user-attachments/assets/5e8f43e9-6e6f-431f-a8dd-fba9e7b96147" />
<img width="1920" height="1080" alt="2025-08-16_18 17 34" src="https://github.com/user-attachments/assets/a1055ea9-6065-4e9a-8160-d536ce2240a4" />

---

The recipes for all the canned foods consisting of a full can of a single item are listed when viewing the usage of the food canner (by default this can be seen by pressing U when hovering over the item). These are the canned foods listed in REI/JEI on the right and also the canned foods creative mode tab.

### REI

<img width="1920" height="1080" alt="2025-08-16_18 15 27" src="https://github.com/user-attachments/assets/c11c488f-3c76-48cd-8266-28254a6a2273" />

### JEI

<img width="1920" height="1080" alt="2026-01-04_01 25 13" src="https://github.com/user-attachments/assets/f5457817-5f97-4e43-bbe6-be34d75b6634" />

---

### REI

https://github.com/user-attachments/assets/df3d76c7-47b8-4d61-8e32-926275dba723

### JEI

https://github.com/user-attachments/assets/2fdc92e2-dcdc-4d86-b747-ba3e8d80faf7

Roelymole committed:
feat: add REI recipe viewer for canning foods

Roelymole committed:
Merge branch 'main' of https://github.com/TeamGalacticraft/Galacticraft into food_canner_rei

Roelymole committed:
feat: list the canning tab in the list of all categories

Roelymole committed:
Merge branch 'main' of https://github.com/TeamGalacticraft/Galacticraft into food_canner_rei

Roelymole committed:
chore: add comment with links for future reference when updating to 1.21.2+

Roelymole committed:
Merge branch 'main' of https://github.com/TeamGalacticraft/Galacticraft into food_canner_rei

Roelymole committed:
chore: update food canner REI integration

Roelymole committed:
chore: dark theme support for the canning REI recipe viewer
Also reduce the use of magic numbers by moving values to Constant.FoodCanner

Roelymole committed:
feat: animate food canner recipe viewer in REI

Roelymole committed:
fix: get transfer handler to work for canning recipe display in REI

Roelymole committed:
feat: replicate food canner recipe viewer in JEI

Roelymole committed:
fix: use the correct value for inventorySlotStart of the food canner JEI transfer handler
Also use FoodCannerMenu.class instead of MachineMenu.class as it is more specific

Roelymole committed:
refactor: move repeated code to FoodCannerProgressAnimation

Roelymole committed:
feat: show canning recipes in EMI for items in the canned food creative tab

Roelymole committed:
chore: update REI/JEI canning recipe viewers

Roelymole committed:
chore: add canning simplified icon

Roelymole committed:
refactor: swap positions of canning and rocket simplified icons

marcus8448 reviewed (changes requested):
A couple bugs:


The transfer handler items seem to be misaligned with the rendered ones (REI):

<img width="886" height="395" alt="image" src="https://github.com/user-attachments/assets/74c0cbb3-9c10-466a-8d9c-8316aecc3b84" />

---

There's an empty can canning recipe (both REI and JEI), if you place an empty can into a canner and view its recipe when it's in the 'processing' slot.
<img width="292" height="189" alt="image" src="https://github.com/user-attachments/assets/f88177f5-58de-48c6-bf6e-5e2faaf9531a" />

---

If you look at the 'uses' of an empty can, the items are placed in the wrong slot (REI).
<img width="324" height="218" alt="image" src="https://github.com/user-attachments/assets/cb3933f5-17da-403c-b1e1-8b6b8543feaa" />

(meanwhile JEI shows no canning recipes when looking at uses for an empty can)

Roelymole committed:
fix: address bugs pointed out in review
Misaligned transfer handler for canning display in REI
Remove empty can recipe viewer in REI and JEI
Prevent food from being shown in the empty can slot for uses in REI
Show canning recipes as uses of the empty can in JEI

marcus8448 commented on src/main/java/dev/galacticraft/mod/compat/jei/category/JEICanningProgressBar.java:
```diff
@@ -0,0 +1,136 @@
+
+        @Override
+        @SuppressWarnings("removal")
+        public List<Component> getTooltip(ItemStack itemStack, TooltipFlag tooltipFlag) {
```
Is there something blocking us from using the non-deprecated version? (non-blocking)

Roelymole replied:
This is why...

`error: JEICanningProgressBar.CurrentCanIngredientRenderer 
is not abstract and does not override abstract method getTooltip(ItemStack,TooltipFlag) in IIngredientRenderer`

It's a bit of an oversight on JEI's side.

Roelymole committed:
refactor: delete outer class attribute and use 'qualified this' expressions instead for canning progress widgets

Roelymole committed:
fix: highlight which slots are missing items correctly

Roelymole committed:
chore: delete the getInputIngredients method which is redundant now that the widgets are in the correct order

marcus8448 reviewed (approved):
Thanks!
